modules/solid_mechanics/src/neml2/userobjects/ExecuteNEML2Model.C

```diff
@@ -0,0 +1,354 @@
+
+  // add user object dependencies by name (the UOs do not need to exist yet for this)
+  for (const auto & uo_name : gather_uo_names)
+    _depend_uo.insert(uo_name);
+
+  for (const auto & output : model().output_axis().variable_accessors(true))
+  {
+    // reserve Batch tensors for each output
+    _outputs[output] = neml2::BatchTensor();
+
+    // reserve Batch tensors for each output derivative
+    for (const auto & uo_name : gather_uo_names)
+      _doutputs[std::make_pair(output, uo_name)] = neml2::BatchTensor();
+  }
+
+  validateModel();
+}
+
+void
+ExecuteNEML2Model::initialSetup()
+{
+  // set of provided and required inputs
+  std::set<neml2::VariableName> _provided_inputs;
+
+  // deal with the time input
+  if (model().input_axis().has_variable(_neml2_time))
+    _provided_inputs.insert(_neml2_time);
+  else if (isParamSetByUser("neml2_time"))
+    paramError("neml2_time", "The provided model has no time input `", _neml2_time, "`.");
+
+  // deal with user object provided inputs
+  for (const auto & uo_name : getParam<std::vector<UserObjectName>>("gather_uos"))
+  {
+    // gather coupled user objects late to ensure they are constructed. Do not add them as
+    // dependencies (that's already done in the constructor).
+    const auto & uo = getUserObjectByName<MOOSEToNEML2>(uo_name, /*is_dependency = */ false);
+    _gather_uos.push_back(&uo);
+
+    const auto uo_var = uo.getNEML2Variable();
+    // check if the model actually has an input corresponding to this UO
+    if (!model().input_axis().has_variable(uo_var))
+    {
+      paramError("gather_uos",
+                 "The supplied user object `",
+                 uo_name,
+                 "` connects to a model input that does not exist (`",
+                 uo_var,
+                 "`).");
+    }
+    _provided_inputs.insert(uo_var);
+  }
+
+  // iterate over set of required inputs and error out if we find one that is not provided
+  std::set<neml2::VariableName> _required_inputs = model().input_axis().variable_accessors(true);
+  for (const auto & input : _required_inputs)
+    if (!_provided_inputs.count(input) && input.start_with("forces"))
+      paramWarning(
+          "gather_uos", "No user object is providing the required model input `", input, "`");
+}
+
+void
+ExecuteNEML2Model::timestepSetup()
+{
+  if (_t_step > 0)
+    advanceStep();
+}
+
+bool
+ExecuteNEML2Model::shouldCompute()
+{
+  // NEML2 computes residual and Jacobian together at EXEC_LINEAR
+  // There is no work to be done at EXEC_NONLINEAR **UNLESS** we are computing the Jacobian for
+  // automatic scaling.
+  if (_fe_problem.computingScalingJacobian())
+    return true;
+
+  if (_fe_problem.currentlyComputingResidualAndJacobian())
+    return true;
+
+  if (_fe_problem.currentlyComputingJacobian())
+    return false;
+
+  return true;
+}
+
+void
+ExecuteNEML2Model::initialize()
+{
+  _elem_to_batch_index.clear();
```

Comment:
Do we need to reconstruct this map _every_ time the UO is executed?

Reply:
That's the safe thing. I naively thought it would make refinement possible, but of course the stored NEML2 old states are not prolonged. Maybe we need to let the MOOSE material system handle old state.

Reply:
Well maybe let's think about this a bit more: Element-to-batch-index depends on a few things:
1. Number of elements
2. Element ID ordering
3. Quadrature rule
4. Insertion of custom quadrature points

Is the above an exhaustive list? If so, I think all of them (the first 3 at least) should trigger onMeshChanged(), and so we can gracefully handle the reconstruction of this map. I am not so sure about number 4, though.

Reply:
I'm not sure we can even do 4. outside of a Dirac kernel. Those QPs do not persist into other systems.

Reply:
I wonder if there is anything else that could change the iteration order over elements. For sure if we have dynamic repartitioning (does that trigger mesh changed?).

Reply:
How about XFEM?